### ISSUE TITLE
Moq1400/Moq1410: MockBehavior analyzers compare boxed enum constants with ==/!= (reference equality) instead of value equality

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1255]
## Summary

An internal code audit found that `SetExplicitMockBehaviorAnalyzer` (**Moq1400**) and `SetStrictMockBehaviorAnalyzer` (**Moq1410**) compare compile-time constants using `==`/`!=` on `object?` operands. Both sides are independently boxed enum underlying values (`IOperation.ConstantValue.Value` and `IFieldSymbol.ConstantValue`), so C# resolves the operator to **reference equality of boxes**, not value equality. The comparisons work today only because Roslyn happens to hand back interned boxes for small integer constants (Moq's `MockBehavior` values are `Strict = 0`, `Loose = 1`, `Default = 1` — verified against both Moq 4.8.2 and 4.18.4 assemblies), which puts every `MockBehavior` constant inside the interned range. That is an unspecified implementation detail of the host Roslyn; if either side ever returns a fresh box, Moq1410 produces a **false positive on code whose behavior is provably `Strict`** (a `const MockBehavior` argument) and the default-value branches become dead (**false negative**). The fix is mechanical: value equality (`Equals`) with an explicit `Optional<object?>.HasValue` check.

## Current behavior

`src/Analyzers/SetExplicitMockBehaviorAnalyzer.cs:45-61`:

```csharp
        IArgumentOperation? mockArgument = arguments.DefaultIfNotSingle(argument => argument.Parameter.IsInstanceOf(mockParameter));

        // Is the behavior set via a default value?
        if (mockArgument?.ArgumentKind == ArgumentKind.DefaultValue && mockArgument.Value.WalkDownConversion().ConstantValue.Value == knownSymbols.MockBehaviorDefault?.ConstantValue)
        {
            TryReportMockBehaviorDiagnostic(context, target, knownSymbols, Rule, DiagnosticEditProperties.EditType.Insert, typeName);
        }

        // NOTE: This logic can't handle indirection (e.g. var x = MockBehavior.Default; new Mock(x);). We can't use the constant value either,
        // as Loose and Default share the same enum value: `1`. Being more accurate I believe requires data flow analysis.
        //
        // The operation specifies a MockBehavior; is it MockBehavior.Default?
        if (mockArgument?.DescendantsAndSelf().OfType<IFieldReferenceOperation>().Any(argument => argument.Member.IsInstanceOf(knownSymbols.MockBehaviorDefault)) == true)
        {
            TryReportMockBehaviorDiagnostic(context, target, knownSymbols, Rule, DiagnosticEditProperties.EditType.Replace, typeName);
        }
```

`src/Analyzers/SetStrictMockBehaviorAnalyzer.cs:70-87`:

```csharp
        IArgumentOperation? mockArgument = arguments.DefaultIfNotSingle(argument => argument.Parameter.IsInstanceOf(mockParameter));

        // Is the behavior set via a default value?
        if (mockArgument?.ArgumentKind == ArgumentKind.DefaultValue && mockArgument.Value.WalkDownConversion().ConstantValue.Value == knownSymbols.MockBehaviorDefault?.ConstantValue
            && TryReportMockBehaviorDiagnostic(context, target, knownSymbols, Rule, DiagnosticEditProperties.EditType.Insert, mockedTypeName))
        {
            return;
        }

        // NOTE: This logic can't handle indirection (e.g. var x = MockBehavior.Default; new Mock(x);)
        //
        // The operation specifies a MockBehavior; is it MockBehavior.Strict?
        if (mockArgument?.Value.WalkDownConversion().ConstantValue.Value != knownSymbols.MockBehaviorStrict?.ConstantValue
            && mockArgument?.DescendantsAndSelf().OfType<IFieldReferenceOperation>().Any(argument => argument.Member.IsInstanceOf(knownSymbols.MockBehaviorStrict)) != true)
        {
            TryReportMockBehaviorDiagnostic(context, target, knownSymbols, Rule, DiagnosticEditProperties.EditType.Replace, mockedTypeName);
        }
```

### Why `==`/`!=` is wrong here

- `mockArgument.Value.WalkDownConversion().ConstantValue` is `Optional<object?>` (`IOperation.ConstantValue`); `.Value` is an `object?` holding a boxed `int` for enum constants. Note the current code never checks `.HasValue` — for a non-constant operand, `.Value` is silently `null`.
- `knownSymbols.MockBehaviorDefault?.ConstantValue` / `MockBehaviorStrict?.ConstantValue` is `IFieldSymbol.ConstantValue`, also `object?` holding a boxed `int`.
- With both operands typed `object?`, C# binds `==`/`!=` to `object.ReferenceEquals` semantics. Two boxed `1`s are only "equal" if they are the *same box*.
- It happens to work today because the current Roslyn hands back interned boxes for common small constants, and `MockBehavior`'s underlying values (0 and 1) are inside that range. Nothing guarantees this across Roslyn versions or symbol/operation kinds.

### Trace: `const MockBehavior B = MockBehavior.Strict;` at SetStrictMockBehaviorAnalyzer.cs:82

```csharp
const MockBehavior B = MockBehavior.Strict;
var mock = new Mock<ISample>(B);
```

- **Today (interned boxes, verified by running the built analyzer against this snippet — no Moq1410 is reported):** the argument's constant is boxed `0`, `MockBehaviorStrict.ConstantValue` is the *same* boxed `0`, so `!=` is `false`, the `&&` short-circuits, and no diagnostic is reported. Correct — the behavior is Strict. Note the second clause would **not** rescue this case: `B` is a reference to the user's constant, so `DescendantsAndSelf()` contains no `IFieldReferenceOperation` for the `MockBehavior.Strict` field itself.
- **With fresh boxes (the latent bug):** `!=` on two distinct boxed `0`s is `true`; the field-reference clause is also `true` (no `MockBehavior.Strict` field reference); Moq1410 **fires on provably-Strict code** (false positive), and its code fix would then rewrite the already-correct argument to `MockBehavior.Strict`.
- **Lines :48 and :73 (`==`)** invert: with fresh boxes they are always `false`, so the `ArgumentKind.DefaultValue` branches become dead (false negative). Nuance the audit did not state: with the current Moq API surface, none of the analyzed entry points (`Mock<T>` constructors, `MockRepository` constructor, `Mock.Of<T>`) declares an *optional* `MockBehavior` parameter, so `ArgumentKind.DefaultValue` is not reachable with genuine Moq today — these two branches are defense for future/other overloads. The reachable, user-visible damage is the `:82` false positive.
- **Bonus latent bug cured by the fix:** at `:48`/`:73`, when `knownSymbols.MockBehaviorDefault` is `null` (doppelganger `MockBehavior` without a `Default` member) *and* the argument has no constant, the comparison is `null == null` → `true`, and the analyzer attempts to report on an arbitrary default-valued argument. The fixed helper returns `false` for a `null`/constant-less field.

### Loose vs Default share the value `1` — the fix must not "improve" matching

Verified from the Moq assemblies: `MockBehavior.Strict = 0`, `MockBehavior.Loose = 1`, `MockBehavior.Default = 1` (`Default` aliases `Loose`). The comment at `SetExplicitMockBehaviorAnalyzer.cs:53-55` is the analyzer's design contract: for Moq1400, *explicit* arguments are matched **by field symbol** (`IsInstanceOf(knownSymbols.MockBehaviorDefault)`), never by constant value, because a constant `1` cannot distinguish an intentional `MockBehavior.Loose` from `MockBehavior.Default`. This issue changes **only the equality mechanics** on the existing constant comparisons; it must **not** extend constant-value matching to explicit arguments (doing so would make Moq1400 flag `MockBehavior.Loose`). The tests below pin this.

## Desired behavior

- All three constant comparisons use value equality: check `Optional<object?>.HasValue`, check the field has a constant (`IFieldSymbol.HasConstantValue`), then `Equals(optional.Value, field.ConstantValue)`.
- No observable behavior change on the current toolchain (the diagnostics reported for every existing test row are identical); the behavior is now guaranteed by the language instead of by box interning.
- Moq1400's field-symbol-based matching for explicit arguments is unchanged.

## Implementation plan

1. **Add a shared helper to `src/Analyzers/MockBehaviorDiagnosticAnalyzerBase.cs`** (place it after `TryHandleMissingMockBehaviorParameter`, currently ending at line 162, and before the `AnalyzeCore` declaration at line 164):

```csharp
    /// <summary>
    /// Determines whether the argument's compile-time constant (after stripping conversions)
    /// equals the constant value of the given enum field, using value equality.
    /// </summary>
    /// <param name="argument">The argument operation to inspect.</param>
    /// <param name="field">The enum field (e.g. <c>MockBehavior.Strict</c>) to compare against.</param>
    /// <returns>
    /// <see langword="true"/> when both the argument and the field have constant values and those
    /// values are equal; otherwise <see langword="false"/>.
    /// </returns>
    /// <remarks>
    /// <see cref="IOperation.ConstantValue"/> and <see cref="IFieldSymbol.ConstantValue"/> are both
    /// boxed <see cref="object"/> values; comparing them with <c>==</c> is reference equality of the
    /// boxes and only works by accident when the host Roslyn happens to intern them. Use
    /// <see cref="object.Equals(object, object)"/> for value equality.
    /// </remarks>
    private protected static bool HasSameConstantValue(IArgumentOperation argument, IFieldSymbol? field)
    {
        if (field is not { HasConstantValue: true })
        {
            return false;
        }

        Optional<object?> constantValue = argument.Value.WalkDownConversion().ConstantValue;
        return constantValue.HasValue && Equals(constantValue.Value, field.ConstantValue);
    }
```

   (`IArgumentOperation` is already in scope via the file's existing `using Microsoft.CodeAnalysis.Operations;`.)

2. **Replace `AnalyzeCore` in `src/Analyzers/SetExplicitMockBehaviorAnalyzer.cs` (lines 29-61) with:**

```csharp
    /// <inheritdoc />
    [SuppressMessage("Design", "MA0051:Method is too long", Justification = "Should be fixed. Ignoring for now to avoid additional churn as part of larger refactor.")]
    private protected override void AnalyzeCore(OperationAnalysisContext context, IMethodSymbol target, ImmutableArray<IArgumentOperation> arguments, MoqKnownSymbols knownSymbols)
    {
        // Extract the type name for the diagnostic message
        string typeName = GetMockedTypeName(context.Operation, target);

        // Check if the target method has a parameter of type MockBehavior
        IParameterSymbol? mockParameter = target.Parameters.DefaultIfNotSingle(parameter => parameter.Type.IsInstanceOf(knownSymbols.MockBehavior));

        // If the target method doesn't have a MockBehavior parameter, check if there's an overload that does
        if (TryHandleMissingMockBehaviorParameter(context, mockParameter, target, knownSymbols, Rule, typeName))
        {
            return;
        }

        IArgumentOperation? mockArgument = arguments.DefaultIfNotSingle(argument => argument.Parameter.IsInstanceOf(mockParameter));

        // Is the behavior set via a default value?
        if (mockArgument?.ArgumentKind == ArgumentKind.DefaultValue && HasSameConstantValue(mockArgument, knownSymbols.MockBehaviorDefault))
        {
            TryReportMockBehaviorDiagnostic(context, target, knownSymbols, Rule, DiagnosticEditProperties.EditType.Insert, typeName);
        }

        // NOTE: This logic can't handle indirection (e.g. var x = MockBehavior.Default; new Mock(x);). We can't use the constant value either,
        // as Loose and Default share the same enum value: `1`. Being more accurate I believe requires data flow analysis.
        //
        // The operation specifies a MockBehavior; is it MockBehavior.Default?
        if (mockArgument?.DescendantsAndSelf().OfType<IFieldReferenceOperation>().Any(argument => argument.Member.IsInstanceOf(knownSymbols.MockBehaviorDefault)) == true)
        {
            TryReportMockBehaviorDiagnostic(context, target, knownSymbols, Rule, DiagnosticEditProperties.EditType.Replace, typeName);
        }
    }
```

3. **Replace `AnalyzeCore` in `src/Analyzers/SetStrictMockBehaviorAnalyzer.cs` (lines 53-87) with:**

```csharp
    /// <inheritdoc />
    [SuppressMessage("Design", "MA0051:Method is too long", Justification = "Should be fixed. Ignoring for now to avoid additional churn as part of larger refactor.")]
    private protected override void AnalyzeCore(OperationAnalysisContext context, IMethodSymbol target, ImmutableArray<IArgumentOperation> arguments, MoqKnownSymbols knownSymbols)
    {
        // Extract the mocked type name for the diagnostic message
        string mockedTypeName = GetMockedTypeName(context.Operation, target);

        // Check if the target method has a parameter of type MockBehavior
        IParameterSymbol? mockParameter = target.Parameters.DefaultIfNotSingle(parameter => parameter.Type.IsInstanceOf(knownSymbols.MockBehavior));

        // If the target method doesn't have a MockBehavior parameter, check if there's an overload that does
        if (TryHandleMissingMockBehaviorParameter(context, mockParameter, target, knownSymbols, Rule, mockedTypeName))
        {
            // Using a method that doesn't accept a MockBehavior parameter, however there's an overload that does
            return;
        }

        IArgumentOperation? mockArgument = arguments.DefaultIfNotSingle(argument => argument.Parameter.IsInstanceOf(mockParameter));

        // Is the behavior set via a default value?
        if (mockArgument?.ArgumentKind == ArgumentKind.DefaultValue && HasSameConstantValue(mockArgument, knownSymbols.MockBehaviorDefault)
            && TryReportMockBehaviorDiagnostic(context, target, knownSymbols, Rule, DiagnosticEditProperties.EditType.Insert, mockedTypeName))
        {
            return;
        }

        // NOTE: This logic can't handle indirection (e.g. var x = MockBehavior.Default; new Mock(x);)
        //
        // The operation specifies a MockBehavior; is it MockBehavior.Strict?
        if (mockArgument is not null
            && !HasSameConstantValue(mockArgument, knownSymbols.MockBehaviorStrict)
            && !mockArgument.DescendantsAndSelf().OfType<IFieldReferenceOperation>().Any(argument => argument.Member.IsInstanceOf(knownSymbols.MockBehaviorStrict)))
        {
            TryReportMockBehaviorDiagnostic(context, target, knownSymbols, Rule, DiagnosticEditProperties.EditType.Replace, mockedTypeName);
        }
    }
```

   **Why the `mockArgument is not null` gate is behavior-preserving:** in the old code, a `null` `mockArgument` made both clauses `true` (`null != boxed0` and `null != true`) and the method then called `TryReportMockBehaviorDiagnostic`, which returns `false` without reporting because `mockArgument` can only be `null` when `mockParameter` is `null`, and in that case `target` has no `MockBehavior` parameter, so `TryGetParameterOfType` fails (`src/Analyzers/MockBehaviorDiagnosticAnalyzerBase.cs:103-106`). The new gate skips that no-op call; no diagnostic is produced either way.

   **Null-field semantics preserved for the Strict clause:** old: if `knownSymbols.MockBehaviorStrict` is `null`, `constant != null` is `true` (for any constant-bearing argument) → proceed. New: `HasSameConstantValue(..., null)` is `false` → `!false` is `true` → proceed. Identical outcome.

## Tests

`tests/Moq.Analyzers.Test/SetExplicitMockBehaviorAnalyzerTests.cs` and `tests/Moq.Analyzers.Test/SetStrictMockBehaviorAnalyzerTests.cs` currently contain no behavior-in-constant or behavior-in-variable rows (a gap the audit's test-quality pass also flagged). All expectations below were **verified by running the built analyzers against these exact snippets** — they pass before *and* after the change on the current toolchain; their purpose is to pin the semantics so a Roslyn box-interning change surfaces as a test failure instead of a shipped FP/FN.

1. Add to the `common` array in `SetExplicitMockBehaviorAnalyzerTests.TestData()` (`tests/Moq.Analyzers.Test/SetExplicitMockBehaviorAnalyzerTests.cs:10-22`; rows are multiplied across 2 namespaces × 2 Moq versions by the existing `.WithNamespaces().WithMoqReferenceAssemblyGroups()` call):

```csharp
            // Behavior supplied via a constant: constant values cannot distinguish MockBehavior.Default
            // from MockBehavior.Loose (both are 1), and the analyzer only matches direct
            // MockBehavior.Default field references (see the NOTE in the analyzer). No diagnostic —
            // pins the documented indirection limitation and guards against constant-value matching
            // ever being applied to explicit arguments (which would flag Loose).
            ["""const MockBehavior B = MockBehavior.Default; new Mock<ISample>(B);"""],
            ["""const MockBehavior B = MockBehavior.Loose; new Mock<ISample>(B);"""],
            ["""const MockBehavior B = MockBehavior.Strict; new Mock<ISample>(B);"""],
            // Behavior supplied via a non-constant variable: not analyzable without data flow. No diagnostic.
            ["""MockBehavior behavior = MockBehavior.Default; new Mock<ISample>(behavior);"""],
```

2. Add to the `common` array in `SetStrictMockBehaviorAnalyzerTests.TestData()` (`tests/Moq.Analyzers.Test/SetStrictMockBehaviorAnalyzerTests.cs:10-22`):

```csharp
            // A constant that folds to MockBehavior.Strict (underlying value 0) must NOT be flagged.
            // This is the case that regresses to a false positive if the constant comparison ever
            // degrades to reference equality of fresh boxes.
            ["""const MockBehavior B = MockBehavior.Strict; new Mock<ISample>(B);"""],
            // A constant that folds to MockBehavior.Loose (1) is not Strict and must be flagged.
            ["""const MockBehavior B = MockBehavior.Loose; {|Moq1410:new Mock<ISample>(B)|};"""],
            // A non-constant variable is not analyzable; the analyzer flags it (documented
            // "can't handle indirection" limitation) — pins current behavior, no regression.
            ["""MockBehavior behavior = MockBehavior.Strict; {|Moq1410:new Mock<ISample>(behavior)|};"""],
```

   Notes: the diagnostic span is the whole creation expression (`context.Operation.CreateDiagnostic`), matching the markup above. The unmarked rows are genuine negative assertions — the test harness fails on any unexpected diagnostic.

3. Run the full existing suites for both analyzers and both code-fix test classes (`SetExplicitMockBehaviorCodeFixTests`, `SetStrictMockBehaviorCodeFixTests`) — all existing rows must pass unchanged.

## Constraints

- `src/Analyzers` targets **netstandard2.0** with Roslyn pinned to **4.8** (ADR-003). `Optional<T>`, `IOperation.ConstantValue`, and `IFieldSymbol.HasConstantValue` are all available there. Do not use newer APIs (e.g. no generic `Enum.IsDefined<T>`-style conveniences).
- **No new dependencies.**
- ADR-001: detection stays symbol-based — this change must not add string/name-based matching, and must not extend constant-value matching to explicit arguments (see the Loose/Default section above).
- Tests are mandatory and must follow repo patterns: `{|Moq1400:...|}`/`{|Moq1410:...|}` markup spans, `[Theory]` + `[MemberData]` with `public static IEnumerable<object[]>` data sources, rows multiplied across 2 namespaces × 2 Moq versions (4.8.2 / 4.18.4) via `.WithNamespaces().WithMoqReferenceAssemblyGroups()`. The negative rows here belong in the per-analyzer markup suites (unmarked rows are genuine negative assertions there); use `AllAnalyzersVerifier` only if adding a standalone cross-analyzer no-diagnostics suite.
- No user-visible behavior change on the current toolchain; therefore no `docs/rules/Moq1400.md`/`Moq1410.md` changes are required.

## Validation

```bash
dotnet format
dotnet build          # must complete with 0 warnings
dotnet test --settings ./build/targets/tests/test.runsettings
```

## Acceptance criteria

- No `==`/`!=` comparisons between `IOperation.ConstantValue.Value` and `IFieldSymbol.ConstantValue` remain in `src/Analyzers/SetExplicitMockBehaviorAnalyzer.cs` or `src/Analyzers/SetStrictMockBehaviorAnalyzer.cs`; all constant comparisons go through the shared `HasSameConstantValue` helper (checks `HasValue`/`HasConstantValue`, then `Equals`).
- New behavior-in-constant/variable rows added to both analyzer test classes exactly as specified, and they pass.
- All pre-existing Moq1400/Moq1410 analyzer and code-fix tests pass unchanged.
- `dotnet build` produces 0 warnings; `dotnet format` is clean.

## Out of scope

- Data-flow analysis to resolve non-constant `MockBehavior` variables (the documented indirection limitation at `SetExplicitMockBehaviorAnalyzer.cs:53-55` stands; the Moq1410 variable row pins it).
- Distinguishing `MockBehavior.Default` from `MockBehavior.Loose` through constants (impossible — same underlying value).
- Consolidating the duplicated `AnalyzeCore` prologues of the two analyzers (a separate maintainability finding, audit item E-4).
- The unused 5-parameter `TryReportMockBehaviorDiagnostic`/`TryHandleMissingMockBehaviorParameter` overloads in the base class (audit item E-8).
